src/query/pipeline/transforms/src/processors/transforms/transform_sort_merge_limit.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright 2021 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::cmp::Reverse;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use common_arrow::arrow::compute::sort::row::RowConverter as ArrowRowConverter;
+use common_arrow::arrow::compute::sort::row::Rows as ArrowRows;
+use common_base::containers::FixedHeap;
+use common_exception::Result;
+use common_expression::types::DataType;
+use common_expression::types::DateType;
+use common_expression::types::NumberDataType;
+use common_expression::types::NumberType;
+use common_expression::types::StringType;
+use common_expression::types::TimestampType;
+use common_expression::with_number_mapped_type;
+use common_expression::DataBlock;
+use common_expression::DataSchemaRef;
+use common_expression::SortColumnDescription;
+use common_pipeline_core::processors::port::InputPort;
+use common_pipeline_core::processors::port::OutputPort;
+use common_pipeline_core::processors::Processor;
+
+use super::sort::Cursor;
+use super::sort::RowConverter;
+use super::sort::Rows;
+use super::sort::SimpleRowConverter;
+use super::sort::SimpleRows;
+use super::AsyncAccumulatingTransform;
+use super::AsyncAccumulatingTransformer;
+
+pub struct TransformSortMergeLimit<R: Rows, Converter> {
+    row_converter: Converter,
+    order_by_cols: Vec<usize>,
+    heap: FixedHeap<Reverse<Cursor<Arc<R>>>>,
+
+    buffer: HashMap<usize, DataBlock>,
+    cur_index: usize,
+}
+
+impl<R, Converter> TransformSortMergeLimit<R, Converter>
+where
+    R: Rows,
+    Converter: RowConverter<R>,
+{
+    pub fn try_create(
+        schema: DataSchemaRef,
+        sort_desc: Vec<SortColumnDescription>,
+        limit: usize,
+    ) -> Result<Self> {
+        let order_by_cols = sort_desc.iter().map(|i| i.offset).collect::<Vec<_>>();
+        let row_converter = Converter::create(sort_desc, schema)?;
+        Ok(TransformSortMergeLimit {
+            row_converter,
+            order_by_cols,
+            heap: FixedHeap::new(limit),
+            buffer: HashMap::with_capacity(limit),
+            cur_index: 0,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl<R, Converter> AsyncAccumulatingTransform for TransformSortMergeLimit<R, Converter>
```

Comment:
Why need `AsyncAccumulatingTransform`?

Reply:
Oh, I should use `AccumulatingTransform`. It's a mistake.